src/box/iproto.cc

```diff
@@ -2224,102 +2331,185 @@ struct iproto_cfg_msg: public cbus_call_msg
 	/** Operation to execute in iproto thread. */
 	enum iproto_cfg_op op;
 	union {
+		/** statistic stucture */
 		struct {
-			/** New URI to bind to. */
-			const char *uri;
-			/** Result address. */
-			struct sockaddr_storage addr;
-			/** Address length. */
-			socklen_t addrlen;
+			size_t mem_used;
+			size_t connections;
+			size_t requests;
 		};
+		/** Pointer to evio_service, used for bind */
+		struct evio_service *binary;
 
 		/** New iproto max message count. */
 		int iproto_msg_max;
 	};
+	struct iproto_thread *iproto_thread;
 };
 
+static inline void
+iproto_thread_fill_binary(struct iproto_thread *iproto_thread,
+			  struct evio_service *binary)
+{
+	strcpy(iproto_thread->binary.host, binary->host);
+	strcpy(iproto_thread->binary.serv, binary->serv);
+	iproto_thread->binary.addrstorage = binary->addrstorage;
+	iproto_thread->binary.addr_len = binary->addr_len;
+	ev_io_set(&iproto_thread->binary.ev, binary->ev.fd, EV_READ);
+}
+
 static inline void
 iproto_cfg_msg_create(struct iproto_cfg_msg *msg, enum iproto_cfg_op op)
 {
 	memset(msg, 0, sizeof(*msg));
 	msg->op = op;
 }
 
+static void
+iproto_fill_stat(struct iproto_thread *iproto_thread,
+		 struct iproto_cfg_msg *cfg_msg)
+{
+	cfg_msg->mem_used =
+		slab_cache_used(&iproto_thread->net_cord.slabc) +
+		slab_cache_used(&iproto_thread->net_slabc);
+	cfg_msg->connections =
+		mempool_count(&iproto_thread->iproto_connection_pool);
+	cfg_msg->requests =
+		mempool_count(&iproto_thread->iproto_msg_pool);
+}
+
 static int
 iproto_do_cfg_f(struct cbus_call_msg *m)
 {
 	struct iproto_cfg_msg *cfg_msg = (struct iproto_cfg_msg *) m;
 	int old;
+	struct iproto_thread *iproto_thread = cfg_msg->iproto_thread;
+
 	try {
 		switch (cfg_msg->op) {
 		case IPROTO_CFG_MSG_MAX:
-			cpipe_set_max_input(&tx_pipe,
+			cpipe_set_max_input(&iproto_thread->tx_pipe,
 					    cfg_msg->iproto_msg_max / 2);
 			old = iproto_msg_max;
 			iproto_msg_max = cfg_msg->iproto_msg_max;
 			if (old < iproto_msg_max)
-				iproto_resume();
+				iproto_resume(iproto_thread);
 			break;
 		case IPROTO_CFG_LISTEN:
-			if (evio_service_is_active(&binary))
-				evio_service_stop(&binary);
-			if (cfg_msg->uri != NULL &&
-			    (evio_service_bind(&binary, cfg_msg->uri) != 0 ||
-			     evio_service_listen(&binary) != 0))
+			if (evio_service_is_active(&iproto_thread->binary)) {
```

Comment:
Why this behaviour changed?  Why can't we still stop listening instead of raising an error?

Reply:
It's because in this patch we use only one bind socket and all iproto threads are listening it. Correct sequence of steps: stop listening socket in all iproto threads, re-bind socket in the main thread, and then start listening socket again in all iproto threads.
In previous version we bind and listen socket in iproto thread so we was not need separate message to stop and to listen.